proposals/extensions.md

```diff
@@ -399,54 +399,194 @@ These requirements need more refinement as implementation progresses, and may ne
+
+The content-based name of the extension grouping type is based on the CLR-level signature and includes the following:
+- The fully qualified CLR name of the type. This will include namespaces, type constraints, named constraints (like `new`).
+    - Type parameter names will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in the fully qualified name being `System.Collections.Generic.Dictionary<T0, T1>`.
+    - The fully qualified name will not include the containing assembly. It is common for types to be moved between assemblies and that should not break the public API.
+- Constraints will included and sorted such that reordering them in source code does not change the name. Specifically:
+    - Type parameter constraints will be listed in declaration order. The constraints for the Nth type parameter will occur before the Nth+1 type parameter.
+    - Base type and interface constraints will be sorted by comparing the full names ordinally
+    - Non-type constraints will be sorted by comparing the C# text ordinally
+- Since this does not include attributes, it intentionally ignores C#-isms like tuple names, nullability, etc ...
+
+Note: The name is guaranteed to remain stable across re-compilations, re-orderings and changes of C#-isms (ie. that don't affect the CLR-level signature).   
+
+#### Extension marker types
+
+The marker type re-declares the type parameters of its containing type (an extension grouping type) to get full fidelity of the C# view of extension blocks.
+
+Each extension block with a distinct C#-level signature in source results in an extension marker type in metadata.  
+- Its name is unspeakable and determined based on the contents of the C#-level signature of the extension block. More details below.  
+- It redeclares the type parameters for its containing type to be those declared in source (including name and attributes).
+- It is private and static.
 - It is marked with the `specialname` flag.  
 
-Method/property declarations in an extension declaration in source are represented as members of the extension type in metadata.  
-The signatures of the original methods are maintained (including attributes), but their bodies are replaced with `throw null`.  
-Those should not be referenced in IL.  
+The content-based name of the extension marker type is based on the following:
+- The fully qualified C# name of the type. 
+    - This will include items like nullable annotations, tuple names, etc ... 
+    - The constraints will have the same ordering as CLR types for the grouping type name
+    - The fully qualified name will not include the containing assembly
+- The name of the extension parameter
+- The modifiers of the extension parameter (`ref`, `ref readonly`, `scoped`, ...)
+- The fully qualified name and attribute arguments for any attributes applied to the extension parameter.
```

Comment:
Attributes and their named arguments should probably be sorted so the content-based name doesn't change after reordering them.